certbot-ci/certbot_integration_tests/utils/misc.py

```diff
@@ -140,13 +140,12 @@ def generate_test_file_hooks(config_dir, hook_probe):
             entrypoint_script = '''\
 #!/usr/bin/env bash
 set -e
-"{0}" "{1}" "{2}" "{3}"
+"{0}" "{1}" "{2}" >> "{3}"
 '''.format(sys.executable, hook_path, entrypoint_script_path, hook_probe)
         else:
-            entrypoint_script_path = os.path.join(hook_dir, 'entrypoint.bat')
+            entrypoint_script_path = os.path.join(hook_dir, 'entrypoint.ps1')
```

Comment:
One more question. I'm not against changing these tests to use Powershell instead of CMD, however, running a `.bat` file should still work right? I tried reverting all the changes to `certbot-ci` and the tests failed though:  https://dev.azure.com/certbot/certbot/_build/results?buildId=1576&view=logs&j=3c6fb28b-7007-5aac-a412-e7418ae4acae&t=8d779bb7-a58f-5e48-61aa-ce0ba6566b9a

Any idea why?

Reply:
Ah yes. This is because CMD (and .bat files) is using the local charset for the Windows region (in my case, in Europe, `CP1252`) while Powershell (and .ps1 files) is using `UTF-16` in every region.

So to have consistent results when running string comparison in the tests, I decided to use only Powershell scripts, and fix the charset to use in assertions to `UTF-16` for Windows.

Reply:
I don't understand. On that branch I reverted all changes to `certbot-ci` so changes to how we parse the probe file were also reverted. Does running the `.bat` file through Powershell change its output to be encoded using `UTF-16`?

Reply:
Well as always with charsets, it is a mess. In the reverted branch you tested, you end up with a python script, writing to a file, which is invoked from a batch file itself called from a Powershell command. It ends up with a file created with a declared charset `UTF-8`, but whose content written by python is in `CP1252`. And so the error when we try to compare this unreadable file with the `deploy` string.

Instead in this PR it is Powershell all along, and so `UTF-16` characters inside a declared `UTF-16` charset file, which can be successfully read by the assertion hook function.

Reply:
Do you understand why this change occurs? My understanding is that in both `master` and in this PR the probe file is created in Python at https://github.com/certbot/certbot/blob/4cce3458f3e6a38fbb3f39613aca1c9e1ebd86ca/certbot-ci/certbot_integration_tests/certbot_tests/context.py#L34

Then in the tests in `master` which I used in the branch above, we call a `.bat` file that runs a Python script that appends to this file. The change to Certbot's production code to run this `.bat` file through Powershell somehow changes the declared encoding of the hook probe? I tried to look into this myself, but ended up wasting time fighting https://github.com/certbot/certbot/issues/7939 and ran out of time for the day.

It's probably fine, but since this is a user facing change that we'll probably have to live with for a long time, I'm trying to understand how running commands through Powershell may change the behavior of commands in ways that sysadmins might not expect as can be seen in the test failures above. While it's certainly OK to change our tests, I see no reason why the tests as they are in `master` shouldn't pass on the production code changes to Certbot in this PR.

Reply:
Ok let's put things that way.

You have 3 possibles ways to write content in a file in the tests, that leads to three difference encoding:
1) using `open("/path/to/file")` in Python: UTF-8
2) builtin CMD write (`output-command >> /path/to/file`): CP1252 (or other depending of the region)
3) builtin Powershell write (`& output-command >> /path/to/file`): UTF-16-LE

In parallel Pyton will assume `UTF-8` in `open` calls that end up reading a file, if not using specific flags or encoding.

On master method 1) is use in the tests when we invoke hooks by files, like it is done in `test_renew_with_hook_scripts`. And method 2) is used each time we have something like `misc.echo('wtf_pre', context.hook_probe)` in parameters for hook flags during a certbot call, since the `echo` method is defined as below:
https://github.com/certbot/certbot/blob/4cce3458f3e6a38fbb3f39613aca1c9e1ebd86ca/certbot-ci/certbot_integration_tests/utils/misc.py#L290-L302

What changes my PR is since hooks commands are not executed with Powershell, the method 3) is effectively used when `echo` function is involved, and the probe files then have `UTF-16-LE` encoding instead of `CP1252`.

At the end, all probe files are asserted with `assert_hook_execution` method, that is invoking `open` and so read all files in `UTF-8`.

Despite having `UTF-8` or `CP1252` encoding probe files on master, things were working well, because `CP1252` is single-byte encoding, and UTF-8 also for ANSI-like characters, which is only what we use here. And `CP1252`, similarly to `UTF-8` had the elegancy to use the same byte value than the ANSI encoding, so at the end everything is correctly read with just assuming everything is `UTF-8`.

However `UTF-16-LE` is a double-byte encoding. Here forget about ANSI encoding retrocompatibility. Then our tests are failing when `assert_hook_execution` tries to read the files written with method 3) assuming `UTF-8` encoding.

There were two possibilities to fix that issue:
- change the logic as I did in this PR, to always have `UTF-16-LE` encoding and assume that when we are on Windows for the tests
- change `echo` function to write probes from Python using `open`, and continue to assume everything is `UTF-8`

I prefered the first solution because I did not want to rely to much on Python to cover the plateform differences, since these integration tests are precisely here to detect errors on specific plateforms. In particular I wanted to be more "real world" with Certbot hooks, as on Windows, everything is embedded and you cannot assume that a Certbot user has in fact a Python distribution installed on his system. Using the builtin shell file writes felt a decent logic to test. Finally the first solution allows to continue to have a minimal logic written in something else than Python.

Reply:
Ah I see! I had forgotten about this `echo` function and thought the`.bat` files (that run Python) from method 1 somehow changed their behavior when ran through Powershell in a way that broke our tests which I was trying to understand as a potential way things might not work as we/our users would expect here.

Your explanation makes sense though and I'm fine with the test changes being made.